File: Python3/413. Arithmetic Slices.py
# 40ms, 67.10%; 12.6MB, 100.00%
class Solution:
    def numberOfArithmeticSlices(self, A: list) -> int:
        diff, count, res = None, 0, 0
        for i in range(1, len(A)):
            tmp = A[i] - A[i - 1]
            if tmp == diff:
                count += 1
            else:
                if count > 2: res += (count - 1) * (count - 2) >> 1
                diff = tmp
                count = 2
        else:
            if count > 2: res += (count - 1) * (count - 2) >> 1
        return res

File: 401 - 600/413 Arithmetic Slices.cc
// 4ms, 63.79%; 8.8MB, 43.75%
class Solution
{
public:
    int numberOfArithmeticSlices(const vector<int> &A)
    {
        if (A.size() < 3) return 0;
        int diff = A[1] - A[0], count = 1, res = 0;
        for (int i = 2, size = A.size(); i < size; ++i) {
            int tmp = A[i] - A[i - 1];
            if (tmp == diff) {
                count += 1;
            } else {
                if (count > 1) res += count * (count - 1) >> 1;
                diff = tmp, count = 1;
            }
        }
        if (count > 1) res += count * (count - 1) >> 1;
        return res;
    }
};
